### PR TITLE
✨Feat(source-sftp): add get file helper for file-transfer

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What

* Bump CDK and Implement the get_file method for file transfer so the connector can provide a better way to get/download a file.
* With the above, we now resolve the @artem1205 comment concern [here](https://github.com/airbytehq/airbyte/pull/47686#discussion_r1821558810).

## How

* Overrirde get_file and implement logic.

## Review guide

1. `airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py`:  use get functionality from the paramiko sftp client instead of rely on file pointer and read method.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py [hunk 1]
```diff
@@ -79,17 +86,63 @@ def get_matching_files(
...
+        logger.info(
+            f"Starting to download the file {file.uri} with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB) "
+            f"to '{local_file_path}' "
+            f"with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB), "
+            f"available disk space: {available_disk_space / (1024 * 1024):,.2f} MB ({available_disk_space / (1024 * 1024 * 1024):.2f} GB),"
+            f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+        )
+        start_download_time = time.time()
+        # Copy a remote file in remote path from the SFTP server to the local host as local path.
+        self.sftp_client.sftp_connection.get(file.uri, local_file_path)
```
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:138]: [get](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L840-L846) method calls [get_fo](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L799-L801) that calls [_transfer_with_callback](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L675-L678) that reads by 32768 bytes chunks.

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py]: * With the above, we now resolve the @artem1205 comment concern [here](https://github.com/airbytehq/airbyte/pull/47686#discussion_r1821558810).

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py [hunk 2]
```diff
@@ -79,17 +86,92 @@ def get_matching_files(
...
+            f"Starting to download the file {file.uri} with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB) "
+            f"to '{local_file_path}' "
+            f"with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB), "
+            f"available disk space: {available_disk_space / (1024 * 1024):,.2f} MB ({available_disk_space / (1024 * 1024 * 1024):.2f} GB),"
+            f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+        )
+        progress_handler = self.create_progress_handler(local_file_path, logger)
+        start_download_time = time.time()
+        # Copy a remote file in remote path from the SFTP server to the local host as local path.
+        self.sftp_client.sftp_connection.get(file.uri, local_file_path, callback=progress_handler)
```
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:167]: [get](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L840-L846) method calls [get_fo](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L799-L801) that calls [_transfer_with_callback](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L675-L678) that reads by 32768 bytes chunks.

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:167]: [get](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L840-L846) method calls [get_fo](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L799-L801) that calls [_transfer_with_callback](https://github.com/paramiko/paramiko/blob/23f92003898b060df0e2b8b1d889455264e63a3e/paramiko/sftp_client.py#L675-L678) that reads by 32768 bytes chunks.

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:167]: With the above, we now resolve the @artem1205 comment concern https://github.com/airbytehq/airbyte/pull/47686#discussion_r1821558810.

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py [hunk 3]
```diff
@@ -79,17 +86,92 @@ def get_matching_files(
...
+                available_memory = memory_info.available
+                logger.info(
+                    f"Available disk space: {available_disk_space / (1024 * 1024):,.2f} MB ({available_disk_space / (1024 * 1024 * 1024):.2f} GB), "
+                    f"available memory: {available_memory / (1024 * 1024):,.2f} MB ({available_memory / (1024 * 1024 * 1024):.2f} GB)."
+                )
+
+        return progress_handler
+
+    @override
+    def get_file(self, file: RemoteFile, local_directory: str, logger: logging.Logger) -> Dict[str, str | int]:
```
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:121]: [5GB >](https://cloud.airbyte.com/workspaces/166f3e04-9021-4f63-a963-a43abb548cec/connections/7e483cf1-3b9a-4e5b-8c4f-c0b74622f5b3/timeline?openLogs=true&eventId=6b7961d3-68d6-41e9-a971-cf39e3362f85) transfer looks good in the testing workspace.

Comment (artem1205) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:121]: Could you please add local integration test to ensure memory usage does not exceed limits? 

P.S. see `pytest.mark.limit_memory`

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/stream_reader.py:121]: @artem1205 tests are now there. Thanks for pointing this!

**Review by artem1205 — APPROVED** [review]
LGTM!

image

Could you please put the `csv` into a `zip` (or any other archive) ? Otherwise, my IDE will literally blow up during indexing.

**Comment by aldogonzalez8** [comment]
> LGTM!
> 
> image
> 
> Could you please put the `csv` into a `zip` (or any other archive) ? Otherwise, my IDE will literally blow up during indexing.

Acknowledge, let me do that.

**Comment by aldogonzalez8** [comment]
> LGTM!
> 
> image
> 
> Could you please put the `csv` into a `zip` (or any other archive) ? Otherwise, my IDE will literally blow up during indexing.

@artem1205 in [this commit ](https://github.com/airbytehq/airbyte/pull/48434/commits/9d8b1342aef47386df3e92105f5ab90907ea4cce)I have placed a base testing file behind a zip, so everything is much lighter now.

**Pull request merged**